File: Solution-iOS/MuslimLib/Source/Models/Utils/Guid.m
//
//  Guid.m
//  MemoreX
//
//  Created by Mohammad Ali Yektaie on 7/29/15.
//  Copyright (c) 2015 Mohammad Ali Yektaie. All rights reserved.
//

#import "Guid.h"

@interface Guid() {
    NSString* _value;
}

@end

@implementation Guid

- (instancetype) initWithString:(NSString*)value {
    self = [super init];
    
    if (self) {
        _value = value;
    }
    
    return self;
}

+ (Guid*)empty {
    return [[Guid alloc] initWithString:@"00000000-0000-0000-0000-000000000000"];
}

- (BOOL)isEmpty {
    return [_value isEqualToString:@"00000000-0000-0000-0000-000000000000"];
}

- (instancetype) initWithByteArray:(ByteArray*)guid {
    self = [super init];
    
    if (self) {
        ByteArray* raw = [[ByteArray alloc] initWithLength:16];
        
        [raw setValue:[guid get:0] atIndex:3];
        [raw setValue:[guid get:1] atIndex:2];
        [raw setValue:[guid get:2] atIndex:1];
        [raw setValue:[guid get:3] atIndex:0];
        [raw setValue:[guid get:4] atIndex:5];
        [raw setValue:[guid get:5] atIndex:4];
        [raw setValue:[guid get:6] atIndex:7];
        [raw setValue:[guid get:7] atIndex:6];
        [raw setValue:[guid get:8] atIndex:8];
        [raw setValue:[guid get:9] atIndex:9];
        [raw setValue:[guid get:10] atIndex:10];
        [raw setValue:[guid get:11] atIndex:11];
        [raw setValue:[guid get:12] atIndex:12];
        [raw setValue:[guid get:13] atIndex:13];
        [raw setValue:[guid get:14] atIndex:14];
        [raw setValue:[guid get:15] atIndex:15];
        
        
        NSString* p1= [self getStringValue:raw start:0 length:8];
        NSString* p2= [self getStringValue:raw start:8 length:4];
        NSString* p3= [self getStringValue:raw start:12 length:4];
        NSString* p4= [self getStringValue:raw start:16 length:4];
        NSString* p5= [self getStringValue:raw start:20 length:12];
        
        _value = p1;
        _value = [_value stringByAppendingString:@"-"];
        _value = [_value stringByAppendingString:p2];
        _value = [_value stringByAppendingString:@"-"];
        _value = [_value stringByAppendingString:p3];
        _value = [_value stringByAppendingString:@"-"];
        _value = [_value stringByAppendingString:p4];
        _value = [_value stringByAppendingString:@"-"];
        _value = [_value stringByAppendingString:p5];
    }
    
    return self;
}

- (NSString*)getStringValue: (ByteArray*)raw start:(int)start length:(int)length {
    NSString* result = [self bytesToHex:raw];
    NSRange range;
    range.location = start;
    range.length = length;
    
    result = [result substringWithRange:range];
    
    return result;
}

- (NSString*) bytesToHex:(ByteArray*)bytes
{
    NSString* hexArray = @"0123456789abcdef";
    NSString* result = @"";
    NSRange range;
    range.length = 1;
    
    for ( int j = 0; j < bytes.length; j++ ) {
        int v = [bytes get:j] & 0xFF;
        
        range.location = v / 16;
        NSString* character = [hexArray substringWithRange:range];
        result = [result stringByAppendingString:character];
        
        range.location = v % 16;
        character = [hexArray substringWithRange:range];
        result = [result stringByAppendingString:character];
    }
    
    return result;
}

+ (Guid*)newGuid {
    NSString* p1 = [self getRandomString:8];
    NSString* p2 = [self getRandomString:4];
    NSString* p3 = [self getRandomString:4];
    NSString* p4 = [self getRandomString:4];
    NSString* p5 = [self getRandomString:12];
    
    NSString* g = p1;
    g = [g stringByAppendingString:@"-"];
    g = [g stringByAppendingString:p2];
    g = [g stringByAppendingString:@"-"];
    g = [g stringByAppendingString:p3];
    g = [g stringByAppendingString:@"-"];
    g = [g stringByAppendingString:p4];
    g = [g stringByAppendingString:@"-"];
    g = [g stringByAppendingString:p5];
    
    return [[Guid alloc] initWithString:g];
}

+ (NSString*)getRandomString:(int)length {
    NSString* hexArray = @"0123456789abcdef";
    NSString* result = @"";
    NSRange range;
    range.length = 1;
    
    for (int i = 0; i < length; i++) {
        range.location = arc4random() % 16;
        NSString* character = [hexArray substringWithRange:range];
        result = [result stringByAppendingString:character];
    }
    
    return result;
}

- (NSString*)description {
    return _value;
}

- (ByteArray*)toByteArray {
    ByteArray* guid = [self toRawArray];
    ByteArray* result = [[ByteArray alloc] initWithLength:16];
    
    [result setValue:[guid get:3] atIndex:0];
    [result setValue:[guid get:2] atIndex:1];
    [result setValue:[guid get:1] atIndex:2];
    [result setValue:[guid get:0] atIndex:3];
    [result setValue:[guid get:5] atIndex:4];
    [result setValue:[guid get:4] atIndex:5];
    [result setValue:[guid get:7] atIndex:6];
    [result setValue:[guid get:6] atIndex:7];
    [result setValue:[guid get:8] atIndex:8];
    [result setValue:[guid get:9] atIndex:9];
    [result setValue:[guid get:10] atIndex:10];
    [result setValue:[guid get:11] atIndex:11];
    [result setValue:[guid get:12] atIndex:12];
    [result setValue:[guid get:13] atIndex:13];
    [result setValue:[guid get:14] atIndex:14];
    [result setValue:[guid get:15] atIndex:15];
    
    return result;
}

- (ByteArray*) toRawArray {
    ByteArray* result = [[ByteArray alloc] initWithLength:16];
    NSString* s = [_value stringByReplacingOccurrencesOfString:@"-" withString:@""];
    NSRange range;
    range.length = 2;
    
    for (int i = 0; i < 16; i++) {
        range.location = i * 2;
        NSString* digit = [s substringWithRange:range];
        
        unsigned hex = 0;
        NSScanner *scanner = [NSScanner scannerWithString:digit];
        
        [scanner setScanLocation:0];
        [scanner scanHexInt:&hex];
        
        int b = (int)hex;
        [result setValue:(char)b atIndex:i];
    }
    
    return result;
}

- (BOOL)isEqual:(id)object
{
    BOOL result = [object isKindOfClass:[Guid class]];
    
    if (result) {
        Guid* g = (Guid*)object;
        result = [[g description] isEqual:[self description]];
    }
    
    return result;
}

@end

File: ContentCreator/src/com/yekisoft/muslim_lib/core/Guid.java
package com.yekisoft.muslim_lib.core;

import java.util.Random;

/**
 * Created by yektaie on 8/5/16.
 */
public class Guid {
    private String _value = null;

    public Guid(String guid) {
        _value = guid;
    }

    public Guid(byte[] guid) {
        byte[] raw = new byte[16];

        raw[3] = guid[0];
        raw[2] = guid[1];
        raw[1] = guid[2];
        raw[0] = guid[3];
        raw[5] = guid[4];
        raw[4] = guid[5];
        raw[7] = guid[6];
        raw[6] = guid[7];
        raw[8] = guid[8];
        raw[9] = guid[9];
        raw[10] = guid[10];
        raw[11] = guid[11];
        raw[12] = guid[12];
        raw[13] = guid[13];
        raw[14] = guid[14];
        raw[15] = guid[15];

        String p1= getStringValue(raw, 0, 8);
        String p2= getStringValue(raw, 8, 4);
        String p3= getStringValue(raw, 12, 4);
        String p4= getStringValue(raw, 16, 4);
        String p5= getStringValue(raw, 20, 12);

        _value = p1 + "-" + p2 + "-" + p3 + "-" + p4 + "-" + p5;
    }

    private String getStringValue(byte[] array, int begin, int length) {
        String result = bytesToHex(array);

        result = result.substring(begin, begin + length);

        return result;
    }

    final protected static char[] hexArray = "0123456789abcdef".toCharArray();
    public static String bytesToHex(byte[] bytes) {
        char[] hexChars = new char[bytes.length * 2];
        for ( int j = 0; j < bytes.length; j++ ) {
            int v = bytes[j] & 0xFF;
            hexChars[j * 2] = hexArray[v >>> 4];
            hexChars[j * 2 + 1] = hexArray[v & 0x0F];
        }
        return new String(hexChars);
    }

    public static Guid NewGuid() {
        Random random = new Random();

        String p1 = getRandomString(8, random);
        String p2 = getRandomString(4, random);
        String p3 = getRandomString(4, random);
        String p4 = getRandomString(4, random);
        String p5 = getRandomString(12, random);

        String g = p1 + "-" + p2 + "-" + p3 + "-" + p4 + "-" + p5;

        return new Guid(g);
    }

    private static String getRandomString(int length, Random random) {
        String chars = "0123456789abcdef";
        String result = "";

        for (int i = 0; i < length; i++) {
            int index = Math.abs(random.nextInt()) % chars.length();
            result += String.valueOf(chars.charAt(index));
        }

        return  result;
    }

    @Override
    public String toString() {
        return _value;
    }

    @Override
    public boolean equals(Object o) {
        boolean same = o instanceof Guid;

        same = same && ((Guid)o)._value.equals(_value);

        return same;
    }

    public byte[] toByteArray() {
        byte[] raw = toRawArray();
        byte[] result = new byte[16];

        result[0] = raw[3];
        result[1] = raw[2];
        result[2] = raw[1];
        result[3] = raw[0];
        result[4] = raw[5];
        result[5] = raw[4];
        result[6] = raw[7];
        result[7] = raw[6];
        result[8] = raw[8];
        result[9] = raw[9];
        result[10] = raw[10];
        result[11] = raw[11];
        result[12] = raw[12];
        result[13] = raw[13];
        result[14] = raw[14];
        result[15] = raw[15];

        return result;
    }

    private byte[] toRawArray() {
        byte[] result = new byte[16];
        String s = _value.replace("-", "");

        for (int i = 0; i < 16; i++) {
            String digit = s.substring(i * 2, i * 2 + 2);
            byte b = (byte)Integer.parseInt(digit, 16);

            result[i] = b;
        }

        return result;
    }

    public static Guid empty() {
        return new Guid("00000000-0000-0000-0000-000000000000");
    }

    public boolean isEmpty() {
        return _value.equals("00000000-0000-0000-0000-000000000000");
    }
}
